### ISSUE TITLE
Object Slicing bug

### ISSUE THREAD

**Issue by ahundt** [jhu-saw/sawConstraintController#14]
It seems lots of objects will be sliced/leak memory because there are parent classes with a virtual destructor and child classes with a nonvirtual destructor.

This applies to the various constraint controller objects that inherit from cmnGenericObject, one particular example (I'm sure there are several) is mtsVFController.

Explanation:
https://stackoverflow.com/questions/461203/when-to-use-virtual-destructors
https://stackoverflow.com/questions/3261694/why-base-class-destructor-virtual-is-called-when-a-derived-class-object-is-del

**Comment by pkazanzides** [comment]
I do not agree. If a base class declares its destructor virtual, then ALL derived class destructors are virtual, whether you include the virtual keyword or not. This is true for any member function, not just destructors. The second stackoverflow link above also states this:  "If the method in base class is marked as virtual all the inherited methods are virtual as well so even if you don't mark the destructors in D and E as virtual they will still be virtual and they still get called in the same order."

**Comment by ahundt** [comment]
Ah didn't know that, thanks! 

I think most of what I said can be ignored, but I'm pretty sure these two warnings are due to actual bugs:
```
/Users/athundt/source/robonetracker/thirdparty/cisst-saw/sawConstraintController/code/mtsVFController.cpp:236:5: Delete called on 'prmKinematicsState' that has virtual functions but non-virtual destructor
/Users/athundt/source/robonetracker/thirdparty/cisst-saw/sawConstraintController/code/mtsVFController.cpp:274:5: Delete called on 'prmSensorState' that has virtual functions but non-virtual destructor
```

**Comment by pkazanzides** [comment]
Yes.  `prmSensorState` is used as a base class for `prmOffsetState`, so its destructor should be virtual.  The same is true for `prmKinematicsState`.

**Comment by pwilken3** [comment]
I agree that these classes should be using virtual destructors. The issue I'm having, though, is that these classes are generated from a cdg file. I'm not aware of any way to override functions created automatically (empty constructor, destructor, etc). I'll continue to investigate, but I haven't seen any example code like this or anything about this in the wiki.

**Comment by ahundt** [comment]
Can't the functions be nonvirtual?

On Sunday, August 9, 2015, Paul Wilkening <notifications@github.com> wrote:

> I agree that these classes should be using virtual destructors. The issue
> I'm having, though, is that these classes are generated from a cdg file.
> I'm not aware of any way to override functions created automatically (empty
> constructor, destructor, etc). I'll continue to investigate, but I haven't
> seen any example code like this or anything about this in the wiki.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jhu-saw/sawConstraintController/issues/14#issuecomment-129235023>
> .
>


-- 

Cheers!
Andrew Hundt

**Comment by pwilken3** [comment]
Sorry for being unclear, I'm specifically referring to overriding non-virtual destructors with virtual destructors.

**Comment by pkazanzides** [comment]
Probably `cisstDataGenerator` needs to be changed to allow generation of virtual destructors. I am assigning this to @adeguet1.
